File: src/GoogleCluster/python/simulator/testThroughput.py
"""
test throughput of database
"""

import os
import csv
import time
import MySQLdb
import threading

# INPUT_PATH = "/media/EEAEDA2FAED9EFD7/GoogleCluster/UsedData/throughput/split_equally_1/"
# INPUT_FILE_LIST = sorted([INPUT_PATH + name for name in os.listdir(INPUT_PATH) if name.endswith('.csv')])
# INPUT_FILE_LIST = ["/media/EEAEDA2FAED9EFD7/GoogleCluster/task_events_one_day/1_1.csv"]
INPUT_FILE_LIST = ["/home/chenzhi/Documents/tmp/1000k.csv"]

INSERT_TASK_EVENTS = '''insert into task_events values (%s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)'''
INSERT_TASK_USAGE = '''insert into task_usage values (%s, %s, %s, %s, %s, %s, %s,
                                                        %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s, %s)'''

def get_current_time():
    localtime = time.localtime()
    time_string = time.strftime("%Y-%m-%d %H:%M:%S", localtime)
    return time_string


def execute(db, cur, row):
    """
    code that will be operated when trigger
    :param row:
    :return: no return
    """
    # logging insert history, but there are too many insert, so just logging fail insert.
    # logging.info('Insert ' + str(time.time()) + ' ' + str(row[0:6]))
    try:
        cur.execute(INSERT_TASK_EVENTS, row)
        db.commit()
    except Exception, e:
        # logging.info('fail to insert value' + str(row))
        # logging.info(traceback.format_exc())
        pass
    db.rollback()


def simulate(db, input_file):
    time.sleep(0.1) # force current thread to release time slice, so other threads can gain time slice
    print threading.currentThread().name, 'start at', get_current_time()

    # you must create a Cursor object. It will let
    #  you execute all the queries you need
    cur = db.cursor()
    reader = csv.reader(input_file, delimiter=',', quoting=csv.QUOTE_NONE)
    row = reader.next()
    try:
        while row is not None:
            execute_time = time.time()
            execute(db, cur, row)
            row = reader.next()
    except StopIteration:
        pass
    input_file.close()
    db.close()
    print 'Insert %s ended.' % threading.current_thread().name

if __name__ == "__main__":
    start_time = time.time()

    thread_pool = []
    # prepare for simulate
    for fileName in INPUT_FILE_LIST:
        db = MySQLdb.connect(host="localhost",  # your host, usually localhost
                          user="root",  # your username
                          passwd="123",  # your password
                          db="googlecluster")  # name of the data base
        open_file = open(fileName, 'r')
        t = threading.Thread(target=simulate, name=fileName, args=(db, open_file,))
        thread_pool.append(t)

    for thread in thread_pool:
        thread.start()


    # wait all threads end.
    for thread in thread_pool:
        thread.join()

    print 'All insert end and cost', str(time.time() - start_time), 's.'

File: src/GoogleCluster/Go/simulator/testThroughput.go
package main

import (
	"database/sql"
	"encoding/csv"
	"fmt"
	_ "github.com/go-sql-driver/mysql"
	"io"
	"io/ioutil"
	"log"
	"os"
	"runtime"
	"strings"
	"sync"
	"time"
)

const (
	INPUT_PATH = "/home/chenzhi/Documents/test/throughput/split_equally_16/"
	QUERY      = "insert into task_events values (?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?, ?)"
)

func currentTimeMilisecond() int64 {
	return time.Time.UnixNano(time.Now()) / 1000 / 1000
}

func simulate(wg *sync.WaitGroup, filename string) {
	fmt.Println(filename, "start at", time.Now())
	csvfile, err := os.Open(filename)
	if err != nil {
		fmt.Println(err)
		return
	}
	reader := csv.NewReader(csvfile)
	db, err := sql.Open("mysql", "root:123@/GoogleCluster")

	for {
		row, err := reader.Read()
		if err != nil {
			if err == io.EOF {
				break
			}
			fmt.Println(err)
			os.Exit(1)
		}
		_, err = db.Exec(QUERY, row[0], row[1], row[2], row[3], row[4], row[5], row[6], row[7], row[8], row[9], row[10], row[11], row[12])
		if err != nil {
			log.Fatal(err)
		}
	}
	db.Close()
	csvfile.Close()
	fmt.Println(filename, "end at", time.Now())
	wg.Done()

}

func main() {
	start_time := currentTimeMilisecond()
	runtime.GOMAXPROCS(runtime.NumCPU())
	var INPUT_FILE_LIST = make([]string, 0)
	contains, _ := ioutil.ReadDir(INPUT_PATH)
	for _, f := range contains {
		if !f.IsDir() && strings.HasSuffix(f.Name(), ".csv") {
			INPUT_FILE_LIST = append(INPUT_FILE_LIST, f.Name())
		}
	}
	wg := sync.WaitGroup{}
	wg.Add(len(INPUT_FILE_LIST))
	for _, name := range INPUT_FILE_LIST {
		go simulate(&wg, INPUT_PATH+name)
	}
	wg.Wait()
	fmt.Println("cost", currentTimeMilisecond()-start_time, "ms")
}
